pkg/verification/verify_test.go

```diff
@@ -469,3 +475,107 @@ func TestVerifyExtendedKeyUsage(t *testing.T) {
 		}
 	}
 }
+
+func createCertChainAndSigner() ([]*x509.Certificate, *signature.ECDSASignerVerifier, error) {
+	sv, _, err := signature.NewECDSASignerVerifier(elliptic.P256(), rand.Reader, crypto.SHA256)
+	if err != nil {
+		return nil, nil, fmt.Errorf("expected NewECDSASignerVerifier to return a signer verifier: %v", err)
+	}
+
+	certChain, err := signer.NewTimestampingCertWithChain(sv)
+	if err != nil {
+		return nil, nil, fmt.Errorf("expected NewTimestampingCertWithChain to return a certificate chain: %v", err)
+	}
+	if len(certChain) != 3 {
+		return nil, nil, fmt.Errorf("expected the certificate chain to have three certificates: %v", err)
+	}
+
+	return certChain, sv, nil
+}
+
+func createSignedTimestamp(certChain []*x509.Certificate, sv *signature.ECDSASignerVerifier) (*timestamp.Timestamp, error) {
+	tsq, err := timestamp.CreateRequest(strings.NewReader("TestRequest"), &timestamp.RequestOptions{
+		Hash:         crypto.SHA256,
+		Certificates: true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unexpectedly failed to create timestamp request: %v", err)
+	}
+
+	req, err := timestamp.ParseRequest([]byte(tsq))
+	if err != nil {
+		return nil, fmt.Errorf("unexpectedly failed to parse timestamp request: %v", err)
+	}
+
+	tsTemplate := timestamp.Timestamp{
+		HashAlgorithm:     req.HashAlgorithm,
+		HashedMessage:     req.HashedMessage,
+		Time:              time.Now(),
+		Policy:            asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 57264, 2},
+		Ordering:          false,
+		Qualified:         false,
+		AddTSACertificate: req.Certificates,
+		ExtraExtensions:   req.Extensions,
+	}
+
+	resp, err := tsTemplate.CreateResponse(certChain[0], sv)
+	if err != nil {
+		return nil, fmt.Errorf("unexpectedly failed to create timestamp response: %v", err)
+	}
+
+	ts, err := timestamp.ParseResponse(resp)
+	if err != nil {
+		return nil, fmt.Errorf("unexpectedly failed to parse timestamp response: %v", err)
+	}
+
+	return ts, nil
+}
+
+func TestVerifyTSRWithChain(t *testing.T) {
+	certChain, sv, err := createCertChainAndSigner()
+	if err != nil {
+		t.Errorf("failed to create certificate chain: %v", err)
+	}
+
+	ts, err := createSignedTimestamp(certChain, sv)
+	if err != nil {
+		t.Errorf("failed to create signed certificate: %v", err)
+	}
+
+	opts := VerifyOpts{
+		Roots:         []*x509.Certificate{certChain[2]},
+		Intermediates: []*x509.Certificate{certChain[1]},
+	}
+
+	err = verifyTSRWithChain(ts, opts)
```

Comment:
One other case that would be good to test is generating a timestamp without the cert (in request, `Certificates: false`), and have the cert in `opts.TSACertificate`, and confirm it verifies

Reply:
All of the above test cases should be included in https://github.com/sigstore/timestamp-authority/pull/181/commits/38a6720aeda5292c589aa040fafa24ded329f6b7